src/chain.h

```diff
@@ -627,10 +644,30 @@ class CDiskBlockIndex : public CBlockIndex
             }
         }
 
-        // ProgPow
-        if (this->nVersion & CBlockHeader::PROGPOW_BLOCK && nTime >= nPowTimeStampActive) {
-            READWRITE(nNonce64);
-            READWRITE(mixHash);
+        if (nTime >= nPowTimeStampActive) {
+            if (fCheckVeilDataHash) {
+                // We want to fail to accept this block. setting it NUll should fail all checks done on it
+                SetNull();
+                return;
+            }
+
+            int nPowType = (nVersion &
+                            (CBlockHeader::PROGPOW_BLOCK | CBlockHeader::RANDOMX_BLOCK | CBlockHeader::SHA256D_BLOCK));
+            switch (nPowType) {
+                case CBlockHeader::PROGPOW_BLOCK:
+                    READWRITE(nNonce64);
+                    READWRITE(mixHash);
+                    break;
+                case CBlockHeader::RANDOMX_BLOCK:
+                    break;
+                case CBlockHeader::SHA256D_BLOCK:
+                    READWRITE(nNonce64);
+                    break;
+                default:
+                    break;
+            }
+
+            READWRITE(hashAccumulators);
         }
     }
```

Comment:
What of the case of an "early" newly mined block?  the case where nTime < PowTimeStamp active, but 
fCheckVeilDataHash != true.

So the above if block:
```
if (nTime >= nPowTimeStampActive) {
}
```
should we not have
```
else {
    // Check for an early new block
    if (!fCheckVeilDataHash)
        SetNull(); return;
}
```

Reply:
@CaveSpectre11 That shouldn't ever happen. As if it does, it will reject the block because it contains a bad version.